opc/resource_lbaas_listener.go

```diff
@@ -0,0 +1,304 @@
+	d.Set("port", result.Port)
+	d.Set("server_pool", result.OriginServerPool)
+	d.Set("server_protocol", result.OriginServerProtocol)
+	d.Set("state", result.State)
+	d.Set("uri", result.URI)
+	d.Set("load_balancer", fmt.Sprintf("%s/%s", lb.Region, lb.Name))
+
+	if err := setStringList(d, "path_prefixes", result.PathPrefixes); err != nil {
+		return err
+	}
+
+	if err := setStringList(d, "policies", result.Policies); err != nil {
+		return err
+	}
+
+	if err := setStringList(d, "certificates", result.SSLCerts); err != nil {
+		return err
+	}
+
+	if err := setStringList(d, "tags", result.Tags); err != nil {
+		return err
+	}
+
+	if err := setStringList(d, "virtual_hosts", result.VirtualHosts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceListenerUpdate(d *schema.ResourceData, meta interface{}) error {
+	lbaasClient, err := meta.(*Client).getLBaaSClient()
+	if err != nil {
+		return err
+	}
+	listenerClient := lbaasClient.ListenerClient()
+	name := getLastNameInPath(d.Id())
+	lb := getLoadBalancerContextFromID(d.Id())
+
+	input := lbaas.UpdateListenerInput{
+		Name: d.Get("name").(string),
+	}
+
+	if balancerProtocol, ok := d.GetOk("balancer_protocol"); ok {
+		input.BalancerProtocol = lbaas.Protocol(balancerProtocol.(string))
+	}
+
+	if enabled, ok := d.GetOk("enabled"); ok {
+		input.Disabled = getDisabledStateKeyword(enabled.(bool))
+	}
+
+	if port, ok := d.GetOk("port"); ok {
+		input.Port = port.(int)
+	}
+
+	if serverProtocol, ok := d.GetOk("server_protocol"); ok {
+		input.OriginServerProtocol = lbaas.Protocol(serverProtocol.(string))
+	}
+
+	if d.HasChange("server_pool") {
+		if _, ok := d.GetOk("server_pool"); ok {
+			// Only the URI Path is need on Update
+			serverPool := updateOrRemoveStringAttribute(d, "server_pool")
+			*serverPool = getURIRequestPath(*serverPool)
+			input.OriginServerPool = serverPool
+		} else {
+			// server pool removed
+			input.OriginServerPool = nil
+		}
+	}
+
+	input.PathPrefixes = updateOrRemoveStringListAttribute(d, "path_prefixes")
+	input.Policies = updateOrRemoveStringListAttribute(d, "policies")
+	input.SSLCerts = updateOrRemoveStringListAttribute(d, "certificates")
+	input.Tags = updateOrRemoveStringListAttribute(d, "tags")
+	input.VirtualHosts = updateOrRemoveStringListAttribute(d, "virtual_hosts")
+
+	result, err := listenerClient.UpdateListener(lb, name, &input)
+	if err != nil {
+		return fmt.Errorf("Error updating Listener: %s", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s/%s", lb.Region, lb.Name, result.Name))
+
+	return resourceListenerRead(d, meta)
+}
+
+func resourceListenerDelete(d *schema.ResourceData, meta interface{}) error {
+	lbaasClient, err := meta.(*Client).getLBaaSClient()
+	if err != nil {
+		return err
+	}
+	listenerClient := lbaasClient.ListenerClient()
+	name := getLastNameInPath(d.Id())
+	lb := getLoadBalancerContextFromID(d.Id())
+
+	if _, err := listenerClient.DeleteListener(lb, name); err != nil {
+		return fmt.Errorf("Error deleting Listener")
```

Comment:
Let's also write out the error when we fail to delete the Listener

Reply:
done